Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,222 @@
+        Task DeleteAsync(string id);
+    }
+
+    public class CustomDeviceModels : ICustomDeviceModels
+    {
+        private const string STORAGE_COLLECTION = "deviceModels";
+        private const string CUSTOMMODEL = "CustomModel";
+
+        private readonly IStorageAdapterClient storage;
+        private readonly ILogger log;
+
+        public CustomDeviceModels(
+            IStorageAdapterClient storage,
+            ILogger logger)
+        {
+            this.storage = storage;
+            this.log = logger;
+        }
+
+        /// <summary>
+        /// Get list of custom device models.
+        /// </summary>
+        public async Task<IEnumerable<DeviceModel>> GetListAsync()
+        {
+            ValueListApiModel data;
+
+            try
+            {
+                data = await this.storage.GetAllAsync(STORAGE_COLLECTION);
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to load device models from storage",
+                    () => new { e.Message, Exception = e });
+                throw new ExternalDependencyException("Unable to load device models from storage", e);
+            }
+
+            try
+            {
+                var results = new List<DeviceModel>();
+                foreach (var item in data.Items)
+                {
+                    var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+                    deviceModel.ETag = item.ETag;
+                    deviceModel.Type = CUSTOMMODEL;
+                    results.Add(deviceModel);
+                }
+
+                return results;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Unable to parse device models loaded from storage",
+                    () => new { e.Message, Exception = e });
+                throw new ExternalDependencyException("Unable to parse device models loaded from storage", e);
+            }
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                this.log.Error("Device model id cannot be empty!", () => { });
+                throw new InvalidInputException("Device model id cannot be empty! ");
+            }
+
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+
+            if (generateId)
```

Comment:
this code path doesn't seem to be tested. Add a test to verify that the ID can be passed or that otherwise the ID is a random string (can do all in the same test if you like)

Reply:
Done.